File: backend/src/main/java/com/taptaza/service/BookingService.java
package com.taptaza.service;

import com.taptaza.dto.request.CreateBookingRequest;
import com.taptaza.dto.response.BookingListResponse;
import com.taptaza.dto.response.BookingResponse;
import com.taptaza.dto.response.PageResponse;
import org.springframework.data.domain.Pageable;

import java.util.UUID;

public interface BookingService {

    /**
     * Create a new booking for the authenticated user.
     *
     * @param userId  the ID of the authenticated user
     * @param request the booking creation request containing company, service, address,
     *                date, time, and additional options
     * @return the created booking response with full details
     */
    BookingResponse createBooking(UUID userId, CreateBookingRequest request);

    /**
     * Get paginated list of bookings for the authenticated user.
     *
     * @param userId   the ID of the authenticated user
     * @param pageable pagination parameters (page number, size, sorting)
     * @return paginated response containing simplified booking list items
     */
    PageResponse<BookingListResponse> getUserBookings(UUID userId, Pageable pageable);

    /**
     * Get detailed information about a specific booking.
     *
     * @param userId    the ID of the authenticated user
     * @param bookingId the ID of the booking to retrieve
     * @return the booking response with full details
     * @throws com.taptaza.exception.ResourceNotFoundException if booking not found or doesn't belong to user
     */
    BookingResponse getBookingById(UUID userId, UUID bookingId);

    /**
     * Cancel a booking for the authenticated user.
     * Only bookings with PENDING or CONFIRMED status can be cancelled.
     *
     * @param userId    the ID of the authenticated user
     * @param bookingId the ID of the booking to cancel
     * @return the updated booking response with CANCELLED status
     * @throws com.taptaza.exception.ResourceNotFoundException if booking not found or doesn't belong to user
     * @throws com.taptaza.exception.BadRequestException       if booking cannot be cancelled
     */
    BookingResponse cancelBooking(UUID userId, UUID bookingId);
}

File: frontend/src/services/api/bookingService.ts
import apiClient from './client';
import {
  CreateBookingRequest,
  BookingResponse,
  PageResponse,
  BookingStatus,
} from './types';

class BookingService {
  /**
   * Create a new booking
   */
  async createBooking(data: CreateBookingRequest): Promise<BookingResponse> {
    const response = await apiClient.post<BookingResponse>('/bookings', data);
    return response.data;
  }

  /**
   * Get user's bookings with optional status filter
   */
  async getMyBookings(params?: {
    page?: number;
    size?: number;
    status?: BookingStatus;
  }): Promise<PageResponse<BookingResponse>> {
    const response = await apiClient.get<PageResponse<BookingResponse>>('/bookings', {
      params: {
        page: params?.page ?? 0,
        size: params?.size ?? 10,
        status: params?.status,
      },
    });
    return response.data;
  }

  /**
   * Get booking by ID
   */
  async getBookingById(id: string): Promise<BookingResponse> {
    const response = await apiClient.get<BookingResponse>(`/bookings/${id}`);
    return response.data;
  }

  /**
   * Cancel a booking
   */
  async cancelBooking(id: string): Promise<BookingResponse> {
    const response = await apiClient.put<BookingResponse>(`/bookings/${id}/cancel`);
    return response.data;
  }

  /**
   * Get upcoming bookings
   */
  async getUpcomingBookings(): Promise<BookingResponse[]> {
    const response = await apiClient.get<BookingResponse[]>('/bookings/upcoming');
    return response.data;
  }

  /**
   * Get past bookings
   */
  async getPastBookings(): Promise<BookingResponse[]> {
    const response = await apiClient.get<BookingResponse[]>('/bookings/history');
    return response.data;
  }
}

export const bookingService = new BookingService();
export default bookingService;
